## Data Analyst Task - Sales Data


### Introduction and Problem Statement

There are *three data sources* given by the sales team:

1. Customer data - xlsx format
2. Order data - csv format
3. Shipping data - json format

**Reporting Requirements** (as given):

1. the total amount spent per country & the pending delivery status for each country.
2. the total number of transactions, total quantity sold, and total amount spent for each customer, along with the product details.
3. the maximum product purchased for each country.
4. the most purchased product based on the age category less than 30 and above 30.
5. the country that had minimum transactions and sales amount.

### EDA of the data sources to verify the accuracy, completeness and reliability of the data sources

In [1]:
# import the required libraries
import pandas as pd
import json

In [2]:
# load the customer excel file
customer = pd.read_excel('/Users/sharathv/Learning/PEI Group Interview Prep/Customer.xls', sheet_name='atkoe-u250m')


In [3]:
# load the order csv file
order = pd.read_csv('/Users/sharathv/Learning/PEI Group Interview Prep/Order.csv')


In [4]:
# load the shippings json file
with open('/Users/sharathv/Learning/PEI Group Interview Prep/Shipping.json', 'r') as f:
    shipping = json.load(f)
    

# convert the json load to Pandas data frame    
shipping = pd.DataFrame(shipping)    

In [5]:
# visual check of the top rows in the data frame
customer.head()

Customer_ID    First     Last  Age Country
0            1   Joseph     Rice   43     USA
1            2     Gary    Moore   71     USA
2            3     John   Walker   44      UK
3            4     Eric   Carter   38      UK
4            5  William  Jackson   58     UAE

In [6]:
# customer data frame - info of number of rows, columns, data types and constraints, if any
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Customer_ID  250 non-null    int64 
 1   First        250 non-null    object
 2   Last         250 non-null    object
 3   Age          250 non-null    int64 
 4   Country      250 non-null    object
dtypes: int64(2), object(3)
memory usage: 9.9+ KB


In [7]:
# check the basic summary statistics of the columns in customer data frame
customer.describe()

Customer_ID         Age
count   250.000000  250.000000
mean    125.500000   47.576000
std      72.312977   18.978011
min       1.000000   18.000000
25%      63.250000   29.000000
50%     125.500000   47.000000
75%     187.750000   63.000000
max     250.000000   80.000000

In [8]:
# check the no. of unique customers in "customer" data frame
customer.Customer_ID.nunique()

250

In [9]:
# visual check of the top rows in the data frame
order.head()

Order_ID      Item  Amount  Customer_ID
0         1  Keyboard     400          139
1         2     Mouse     300          250
2         3   Monitor   12000          239
3         4  Keyboard     400          153
4         5  Mousepad     250          153

In [10]:
# order data frame - info of number of rows, columns, data types and constraints, if any
order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Order_ID     250 non-null    int64 
 1   Item         250 non-null    object
 2   Amount       250 non-null    int64 
 3   Customer_ID  250 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 7.9+ KB


In [11]:
# check the basic summary statistics of the columns in order data frame
order.describe()

Order_ID       Amount  Customer_ID
count  250.000000    250.00000   250.000000
mean   125.500000   2130.00000   130.404000
std     72.312977   3575.43493    69.192711
min      1.000000    200.00000     4.000000
25%     63.250000    300.00000    71.500000
50%    125.500000    400.00000   125.500000
75%    187.750000   1500.00000   190.750000
max    250.000000  12000.00000   250.000000

In [12]:
# check the no. of unique customers in the order data frame
order.Customer_ID.nunique()

160

In [13]:
# check the no. of unique Order_IDs
order.Order_ID.nunique()

250

In [14]:
# visual check of the top rows in the data frame
shipping.head()

Shipping_ID     Status  Customer_ID
0            1    Pending          173
1            2    Pending          155
2            3  Delivered          242
3            4    Pending          223
4            5  Delivered           72

In [15]:
# check if a customer has multiple shipping IDs
shipping.Shipping_ID.nunique()
shipping.Customer_ID.nunique()

154

In [16]:
# shipping data frame - info of number of rows, columns, data types and constraints, if any
shipping.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Shipping_ID  250 non-null    int64 
 1   Status       250 non-null    object
 2   Customer_ID  250 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 6.0+ KB


In [17]:
# check the basic summary statistics of the columns in shipping data frame
shipping.describe()

Shipping_ID  Customer_ID
count   250.000000   250.000000
mean    125.500000   120.620000
std      72.312977    73.893848
min       1.000000     1.000000
25%      63.250000    53.250000
50%     125.500000   118.000000
75%     187.750000   187.500000
max     250.000000   248.000000

In [18]:
# check the unique values present in the "status" column of the data frame
shipping.Status.unique()

array(['Pending', 'Delivered'], dtype=object)

In [19]:
# check the no. of unique customers in the shipping data frame 
shipping.Customer_ID.nunique()

154

In [20]:
shipping_duplicates = shipping['Shipping_ID'].duplicated().any()

In [21]:
# check if there are any shipping_id duplicates 
shipping_duplicates

False

In [22]:
shipping_cus_duplicates = shipping['Customer_ID'].duplicated().any()

In [23]:
# check if there are any customers with multiple entries. 
# this could be possible if the customer has multiple orders. 
shipping_cus_duplicates

True

In [24]:
# join (inner) customer and order data frames using Customer_ID as the joining key

cus_order = pd.merge(order, customer, on='Customer_ID', how='inner')

In [25]:
# visual check of the top rows of the merged data frame
cus_order.head()

Order_ID      Item  Amount  Customer_ID    First    Last  Age Country
0         1  Keyboard     400          139     Ryan  Martin   61      UK
1         2     Mouse     300          250  Stephen   Jones   22     USA
2         3   Monitor   12000          239    Janet  Holmes   36      UK
3       211  Keyboard     400          239    Janet  Holmes   36      UK
4         4  Keyboard     400          153    Janet  Valdez   29      UK

In [26]:
cus_order.shape

(250, 8)

In [27]:
# check the no. of unique customers in the combined data frame
cus_order.Customer_ID.nunique()

160

In [28]:
# get the list of customers who have more than one order in the cus_order data frame
cus_order_counts = cus_order['Customer_ID'].value_counts()

In [29]:
# create a data frame of customers with single orders
# we will use this list of customers to check if they have multiple shipping entries
single_order_customers = sorted(cus_order_counts[cus_order_counts==1].index.tolist())

In [30]:
# convert to a data frame
single_order_customers = pd.DataFrame(single_order_customers)
single_order_customers.columns = ['Customer_ID']

In [31]:
# print the list of single order customers
single_order_customers

Customer_ID
0             4
1             5
2            10
3            12
4            15
..          ...
95          242
96          243
97          244
98          246
99          250

[100 rows x 1 columns]

In [32]:
# get the shipping data for these single order customers
single_order_cus_shipping = shipping[shipping['Customer_ID'].isin(single_order_customers['Customer_ID'])]

In [33]:
single_order_cus_shipping

Shipping_ID     Status  Customer_ID
2              3  Delivered          242
5              6    Pending           29
6              7    Pending          236
9             10  Delivered           17
11            12    Pending           37
..           ...        ...          ...
239          240  Delivered           10
242          243  Delivered          135
245          246    Pending          171
246          247    Pending           40
247          248  Delivered           37

[97 rows x 3 columns]

In [34]:
# check for duplicate shipping records for these single order customers. we take customer_id as the counting key
single_order_cus_shipping_duplicates = single_order_cus_shipping['Customer_ID'].value_counts()

In [35]:
single_order_cus_shipping_duplicates

35     4
242    3
40     3
91     3
30     3
      ..
196    1
218    1
228    1
150    1
188    1
Name: Customer_ID, Length: 61, dtype: int64

In [36]:
# taking one of the customers who had one order but multiple shipping_ids
shipping_duplicate = shipping[shipping['Customer_ID']==35]

In [37]:
shipping_duplicate

Shipping_ID     Status  Customer_ID
55            56    Pending           35
101          102    Pending           35
145          146    Pending           35
179          180  Delivered           35

In [38]:
# cross-check to see if this customer had indeed one order in order data frame
cus_35_order = order[order['Customer_ID']==35]

In [39]:
cus_35_order

Order_ID      Item  Amount  Customer_ID
120       121  Keyboard     400           35

There seems to be multiple shipping records for one single order

In [40]:
# join cus_order data frame with shipping info which is available at a customer level. Join using Customer_ID

cus_order_ship = pd.merge(cus_order, shipping, on='Customer_ID', how='inner')

In [41]:
cus_order_ship.shape

(233, 10)

In [42]:
# visual check of the top rows in the combined data frame
cus_order_ship.head()

Order_ID      Item  Amount  Customer_ID   First     Last  Age Country  \
0         6  Harddisk    5000          164  Sandra  Mcmahon   34      UK   
1        59   DDR RAM    1500          164  Sandra  Mcmahon   34      UK   
2         8  Mousepad     200           98   Steve    Braun   72      UK   
3       133   Monitor   12000           98   Steve    Braun   72      UK   
4       202     Mouse     300           98   Steve    Braun   72      UK   

   Shipping_ID     Status  
0          220  Delivered  
1          220  Delivered  
2          122    Pending  
3          122    Pending  
4          122    Pending

In [43]:
# check the no. of unique customers in the combined data frame
cus_order_ship.Customer_ID.nunique()

99

In [44]:
cus_order_ship_counts = cus_order_ship['Customer_ID'].value_counts()

In [45]:
multi_order_customers_ship = sorted(cus_order_ship_counts[cus_order_ship_counts>1].index.tolist())

In [46]:
len(multi_order_customers_ship)

63

In [47]:
# check for duplicates in shipping data frame. We can check the column "Shipping_ID" for the same
duplicate_shipping_ids = shipping['Shipping_ID'].duplicated()

In [48]:
duplicate_shipping_ids = shipping[shipping['Shipping_ID'].duplicated(keep=False)]

In [49]:
duplicate_shipping_ids

Empty DataFrame
Columns: [Shipping_ID, Status, Customer_ID]
Index: []

 No duplicate shipping ids are found

### Summary of the findings 
* Customer data has in total **250** unique customers.
* Order data has in total **250** orders from **160** customers. There are records where a customer has made multiple orders and a few customers who have not made any orders at all. Therefore, we can deduce that the *customer to order must have a one to many cardinality*.
* Shipping data has in total **154** customers data. The status can be either "**Pending**" or "**Delivered**" based on the given data set. Since the data available is minimal, there are only two statuses as of now but there can be more statuses in the complete data set.  
* There are **duplicate entries** in the **shipping data** for customers with single orders. 
* Shipping data provided is **on an individual Order_ID level** but there is **no Shipping_ID in the Order table** to track order for which the shipping status is valid for. Hence for customers with multiple orders, it is not possible to map the shipping status with the appropriate order. This is a major flaw in the data flow especially when there is a need to report the data on an order level to see which orders are not fulfilled. 
* A single "**Order_ID**" must have only one shipping status in the shipping data. However, there are instances where there are multiple entries for a single order. For example, Customer_ID = 35. 
* The **quantity of the item ordered** is not available in the dataset. Ideally, this info should have been available in the orders data. This is another major gap in the data from the report requirements perspective. 
* Shipping data has a column named "**Shipping_ID**" which is not available in the orders table. This point is a further observation of the previous point. 
* By "**product details**", I presume it is the item description. If more details of the product is required from the reporting perspective, we need the data for products in detail. To be checked with the Sales Team! 
* Upon joining all the three data frames (orders, customer & shipping) to get the list of customers who have the complete information, we have in total just **99 customers**.



### Requirements for the anticipated datasets

* The data is **incomplete** for **orders and shipping**. There is no order data for 90 customers and there is no shipping data for 6 customers for whom we have the order data -> *<ins>We need orders and shipping data for all the customers</ins>*.


* Consequently, **Order** data needs to have the "**Shipping_ID**" since one customer can have multiple orders and it cannot be assumed that all orders are fulfilled at the same time. Current shipping info at customer level is inadequate and not reliable to use -> *<ins> "Shipping_ID" needs to be added to Orders table </ins>*.

* "**Item Quantity**" info is missing in the Orders data. This information is required from the reporting perspective to report "**Total Quantity Sold**" -> *<ins> "Item Quantity" field is required in the Order table.</ins>*

* "**Order_Date**" is missing. This is one of the basic fields and there is a lot of insights that can be obtained from time period such as the customer's buying pattern, seasonal trend, and helps with month-on-month, quarter-on-quarter and other periodic analysis -> *<ins> "Order_Date" column is required in the Order table </ins>*.





**Note**: EDA also was done visually, both in Power BI with quick visuals and Excel in some cases. Although the data size was very limited in this case, in general, the visual approach would work as a complementary step to the programmatic way of doing the EDA (SQL, R or Python)